changelog/dmd.identifier-quick-check.dd

```diff
@@ -0,0 +1,14 @@
+Identifiers are verifiable against normalization form C
+
+It is now possible to verify identifiers against normalization form C, as per UAX31 and C23.
+
+The quick check algorithm that is in use is an incomplete form that does not handle maybes and can result in false positives in errors.
+The purpose of this algorithm in its incomplete state is to minimize normalization in a future update.
+
+It is configurable for both D and ImportC using ``-normalization=<strategy>`` and ``-normalization-importc=<strategy>``.
+Using the options ``ignore``, ``deprecate``, ``warn``, and ``error``.
+
+Unless you use non-ASCII characters, this change will not affect you.
+
+Due to bug [24507](https://issues.dlang.org/show_bug.cgi?id=24507) if you use GCC as your preprocessor with any identifier that is not normalized to form C, by default it will produce a warning.
```

Comment:
Can't we just support C23 and leave it at that?

Reply:
We are, the problem is GCC is picking different behaviors on failure that we want to override.

Reply:
As for the rest of the text, normalization form C is the big breaking change that people need to be aware of in this PR.

Which is what C23 is defined against.

> D.5 Equivalent Normalized Identifiers
1 UAX #31 requires that implementations describe how identifiers are compared and considered
equivalent.
2 C requires that identifiers be in Normalization Form C and therefore identifiers that compare the
same under NFC are equivalent. This is described in 6.4.2.

Reply:
I'd be surprised if anyone even uses normalized identifiers. If we support various incompatible normalized identifier schemes, then the user will be creating code that won't link with other ImportC code.

I suggest pick the C23 scheme and not do any of these other bug accommodations.

Reply:
Does gcc have some flag to pick different schemes?

Reply:
Yes. They can pick between a few, but the important ones are [C and KC](https://gcc.gnu.org/onlinedocs/gcc/Warning-Options.html#index-Wnormalized_003d).

I do **_not_** want to offer any other normalization forms than C to a user.
D form is easily done once you've got C (it's a subset) so I'll write the wrapper function for it during the next PR that adds the normalization function itself.
But the KC/KD require extra tables, while I'm not against seeing them added I don't see a reason for dmd to need it.

The compiler switches introduced here are to pick what to do when it is not normalized.
Right now that is to accept it as-is (with varying levels of errorness) and later that would be to normalize it.
Without this, I cannot verify the quick check algorithm as part of the test suite currently.

The goal of the switches is to give a ± 10 release window when people can transition forward and backward as required.
That window can be changed, but I believe that may work well so as to try and get one GDC release on either side of changing defaults to C23.

We may not keep any of these switches long-term.
Before removing them happens ImportC should be defined against C23 and we need to transition everyone over to C23.
Once that happens picking tables becomes unnecessary and can all be cleaned up.

But until then to switch 100% over right now would be a breaking change without a transition backwards hence the need for switches.
Unless of course, you're telling me to do the hard break, at which point I'll do a different PR removing all that infrastructure.

Iain may have opinions on switches, as well as if we support NFKC. So we do need to hear from him.